File: app/src/main/java/com/hxh/kt10/java/LazyThreadSafeSynchronized.java
package com.hxh.kt10.java;

/**
 * Created by HXH on 2017/6/1 0001.
 * 懒加载单例(线程安全的-方法同步)
 */

public class LazyThreadSafeSynchronized {
    private static LazyThreadSafeSynchronized INSTANCE;

    private LazyThreadSafeSynchronized() {

    }

    public static synchronized LazyThreadSafeSynchronized getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new LazyThreadSafeSynchronized();
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/hxh/kt10/kotlin/LazyThreadSafeSynchronized.kt
package com.hxh.kt10.kotlin

/**
 * Created by HXH on 2017/6/1 0001.
 *
 */
class LazyThreadSafeSynchronized private constructor() {
    companion object {
        private var instance: LazyThreadSafeSynchronized? = null

        fun get(): LazyThreadSafeSynchronized {
            if (instance == null) instance = LazyThreadSafeSynchronized()
            return instance!!
        }
    }
}

fun main(args: Array<String>) {
    LazyThreadSafeSynchronized.get().sayHello()
}

fun LazyThreadSafeSynchronized.sayHello() {
    println("hello")
}